ext/io/utils.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::futures::task::AtomicWaker;
+use deno_core::futures::Future;
+use deno_core::parking_lot::Mutex;
+use std::collections::LinkedList;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+#[derive(Default)]
+struct TaskQueueTaskWaker {
+  is_ready: AtomicBool,
+  waker: AtomicWaker,
+}
+
+#[derive(Default)]
+struct TaskQueueTasks {
+  is_running: bool,
+  wakers: LinkedList<Arc<TaskQueueTaskWaker>>,
+}
+
+/// A queue that executes tasks sequentially one after the other
+/// ensuring order and that no task runs at the same time as another.
+///
+/// Note that tokio's semaphore doesn't seem to maintain order
+/// and so we can't use that in the code that uses this or use
+/// that here.
+#[derive(Clone, Default)]
+pub struct TaskQueue {
```

Comment:
This just reminded me that I wrote this code before and we should use this here: https://github.com/denoland/deno/pull/18014

This code should not be "io" specific, so we should move it elsewhere.

Reply:
Opened https://github.com/denoland/deno/pull/18016